lucene/core/src/java/org/apache/lucene/search/Matches.java

```diff
@@ -44,4 +47,9 @@
    */
   Collection<Matches> getSubMatches();
 
+  /**
+   * Find all matching terms
+   */
+  void getMatchingTerms(Consumer<Term> termsConsumer) throws IOException;
+
```

Comment:
Wouldn't this make more sense as a method of `MatchesIterator`?

In conjunction with some changes to `PhraseWeight` and `SloppyPhraseMatcher`, that would allow us to accurately capture reordered phrases, which I think `getMatchingTerms` would be a good API for.

My proposed use case is to modify `AnalyzingInfixSuggester` to use a `PhraseQuery` (with slop) and expose this data in order to suggest a reordered phrase to the user without including the entire document in the response.

Reply:
Can you not already do this with the plain `MatchesIterator`?  It will give you the start offset and end offset of matching sloppy phrases, and you can use those to pull phrases from the source.